cache.go

```diff
@@ -0,0 +1,110 @@
+package nds
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+)
+
+// MultiError is returned by batch operations when there are errors with
+// particular elements. Errors will be in a one-to-one correspondence with
+// the input elements; successful elements will have a nil entry.
+type MultiError []error
+
+func (m MultiError) Error() string {
+	s, n := "", 0
+	for _, e := range m {
+		if e != nil {
+			if n == 0 {
+				s = e.Error()
+			}
+			n++
+		}
+	}
+	switch n {
+	case 0:
+		return "(0 errors)"
+	case 1:
+		return s
+	case 2:
+		return s + " (and 1 other error)"
+	}
+	return fmt.Sprintf("%s (and %d other errors)", s, n-1)
+}
+
+var (
+	// ErrCacheMiss means an item was not found in cache
+	ErrCacheMiss = errors.New("nds: cache miss")
+	// ErrCASConflict means the cache item was modified between Get and CAS calls
+	ErrCASConflict = errors.New("nds: cas conflict")
+	// ErrNotStored means that an item was not stored due to a condition check failure (e.g. during an Add or CompareAndSwap call)
+	ErrNotStored = errors.New("nds: not stored")
+)
+
+// Cacher represents a cache backend that can be used by nds.
+type Cacher interface {
+	// NewContext returns the same or a derivative of the passed in context. The returned context will
+	// be used in subsequent calls to the other Cacher functions.
+	NewContext(c context.Context) (context.Context, error)
+	// AddMulti adds each provided Item into the cache if and only if the key for the item is not
+	// currently in use. For any item that was not stored due to a key conflict, a MultiError is returned
+	// with ErrNotStored in the corresponding index for that item. There may be other errors for each MultiError index.
+	AddMulti(c context.Context, items []*Item) error
```

Comment:
A general rule of thumb, at least in my circles, is to use slices of values (`[]Item`) instead of slices of pointers (`[]*Item`). This is because the whole slice and items within it gets allocated in one go which is much more efficient on the allocator. (Note I have just used this method call as an example.)

Reply:
This is just an artifact from basing my implementation off of AppEngine's [memcache.AddMulti](https://github.com/golang/appengine/blob/master/memcache/memcache.go#L333)

I could switch it since we're somewhat breaking our reliance on a specific implementation, I agree  in terms of efficient allocations.

Reply:
I think I'm going to leave this as is, slice of pointers is used through the [datastore library](https://godoc.org/cloud.google.com/go/datastore#Client.DeleteMulti) and other examples I looked at.

Reply:
FYI - https://stackoverflow.com/questions/27622083/performance-slices-of-structs-vs-slices-of-pointers-to-structs

When manipulating structs, it looks like slice of pointers wins. There is at least one spot in the code where we build a slice based on what we were able to cache.

I tried a similar benchmark using the `Item` struct with non-arbitrary data with Go 1.11.2 and using a slice of pointers is marginally faster:

> BenchmarkAppendingStructs-16            20000000               109 ns/op
> BenchmarkAppendingPointers-16           20000000                84.7 ns/op

Reply:
link to gist: https://gist.github.com/someone1/a317313bea260546c58b45b4ea811125

Reply:
Thanks for the benchmarks. I'm happy with your choice and pleasantly surprised.